modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ReplaceableConstDoubleValueSource.java

```diff
@@ -15,20 +15,21 @@
 import org.apache.lucene.search.IndexSearcher;
 
 import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 /**
  * A {@link DoubleValuesSource} which has a stub {@link DoubleValues} that holds a dynamically replaceable constant double.
  */
 final class ReplaceableConstDoubleValueSource extends DoubleValuesSource {
-    final ReplaceableConstDoubleValues fv;
 
-    ReplaceableConstDoubleValueSource() {
-        fv = new ReplaceableConstDoubleValues();
-    }
+    private final Map<LeafReaderContext, ReplaceableConstDoubleValues> specialValues = new ConcurrentHashMap<>();
```

Comment:
I am not entirely happy with this solution, but given that we need a global data structure exposed in the bindings, which will retrieve values per segment, I think we don't have many other options than tracking this special value per segment. This needs to be a concurrent hash map because getValues and setValue are called concurrently from different threads (one per slice)

Reply:
I'm also uneasy, and I think the longer-term solution here is that we need to build the aggregations per-slice rather than globally.  But that's a much bigger change, so I think I'm happy with this as a quick fix.

Reply:
I wonder if this will significantly hurt performance on expression score scripts which also use this class.

Reply:
I don't think it will, because it would only try to access the map if using the value source, which will only occur if `_value` appears in the script script, which is (I think) bogus.

Reply:
Changes for aggregations too right?

Reply:
_value is specifically part of aggregation scripts. But I do not believe explain would ever be called on an it in that context.